Add Ability to Hide Unwanted Page Elements

ZemyCode opened the pull request:
## Hide Elements Feature

This feature addition adds a new menu in the settings for hiding certain page elements based on requests from different issues.

This is what the new menu looks like: 

![image](https://user-images.githubusercontent.com/69816894/235015056-56bfcc75-6682-4a26-8389-73bcbd0d65e0.png)

Various settings added with false being default value for all of them:

![image](https://user-images.githubusercontent.com/69816894/235015139-0248fce4-5ce5-4d56-b225-e32d1cf81689.png)
![image](https://user-images.githubusercontent.com/69816894/230739177-8e64ac97-0592-48a6-9675-829cc0e34cb5.png)
![image](https://user-images.githubusercontent.com/69816894/230739182-06b6e05a-68cc-4924-8c2f-88a10e4bd75f.png)

Feedback will be wanted for the names and descriptions of the settings.

Also requesting feedback on the performance or code quality as I am not sure if calling "updateHiddenElements()" on every setting change is optimal.

**Edited 4/27/2023:** Removed "Home Page" feature for hiding Front Page Carousel due to being buggy, may look into this more in the future.

Melonify commented:
Maybe hiding offline channels in the sidebar should be an option here as well.

Also I think the settings tab should be named something more generic, like Layout or something, or be a sub-section of Appearance -> Interface

ZemyCode commented:
> Maybe hiding offline channels in the sidebar should be an option here as well.
> 
> Also I think the settings tab should be named something more generic, like Layout or something, or be a sub-section of Appearance -> Interface

@Melonify I looked into making it a subsection of Appearance -> Interface but it does not seem a subsection of a subsection is supposed in the code right now. I think there are too many "Hide" settings for all of them to just be a part of the "Interface" subsection, should I just go with the "Layout" idea? 

Or should I make a subsection of Appearance called "Hide" or "Hide elements"? The issue with this one would be I feel like having subsections for "Hide" is important like the current subsections I already made called "Chat", "Side bar" etc. since it makes it easier for the user to find the specific element they want to hide.

Melonify commented:
Maybe a full section is better, perhaps "Site Layout", is a better name than "Layout"

AnatoleAM commented:
@ZemyCode any updates on this PR?

ZemyCode commented:
> @ZemyCode any updates on this PR?

@AnatoleAM I have been out of town and unable to work on this but can work on it soon. Iirc the last thing I have to do is edit how it appears in the menu and just test things a little more.

ZemyCode commented:
@AnatoleAM I think this can be reviewed now, changed the menu from "Hide" to "Site Layout", removed the hide front page carousel since it was buggy and I may look into adding it in a future PR. Also included the "gift a sub" button in the "Hide Subscribe Button" setting.

AnatoleAM commented on src/site/twitch.tv/modules/hidden-elements/HiddenElementsModule.vue:
```diff
@@ -0,0 +1,233 @@
+
+.seventv-hide-react-buttons {
+	div[class$="theatre-social-panel"]
+		button[class="ScCoreButton-sc-ocjdkq-0 ScCoreButtonSecondary-sc-ocjdkq-2 ibtYyW bTKXKk"],
+	div[data-target="channel-header-right"]
+		button[class="ScCoreButton-sc-ocjdkq-0 ScCoreButtonText-sc-ocjdkq-3 ibtYyW jYfhUy"] {
```
The classes here are generated and will change anytime Twitch deploys a new version of their site.

AnatoleAM commented on src/site/twitch.tv/modules/hidden-elements/HiddenElementsModule.vue:
```diff
@@ -0,0 +1,233 @@
+}
+
+.seventv-hide-community-highlights {
+	div[class="Layout-sc-1xcs6mc-0 kAIqwe"] {
```
The classes here are generated and will change anytime Twitch deploys a new version of their site.

AnatoleAM commented on src/site/twitch.tv/modules/hidden-elements/HiddenElementsModule.vue:
```diff
@@ -0,0 +1,233 @@
+		() => {
+			updateHiddenElements();
+		},
+		{ deep: true },
```
Avoid `{ deep: true }` as it leads to major performance issues

AnatoleAM commented on src/site/twitch.tv/modules/hidden-elements/HiddenElementsModule.vue:
```diff
@@ -0,0 +1,233 @@
+	});
+}
+
+onMounted(() => {
+	updateHiddenElements();
+
+	watch(
+		() => hiddenElementSettings,
+		() => {
+			updateHiddenElements();
+		},
```
This would be better using `watchEffect()` instead of `onMounted()`. The inner watch and outer function is unnecessary

ZemyCode replied:
Thanks for this suggestion, still learning about Vue reactivity so I thought there would be a better way. watchEffect works perfectly.

AnatoleAM commented on src/site/twitch.tv/modules/hidden-elements/HiddenElementsModule.vue:
```diff
@@ -0,0 +1,233 @@
+/**
+ * Settings configuration for hidden elements feature
+ */
+export const config = [
```
Rename key prefix to `layout.hide`

AnatoleAM reviewed (approved):
Thanks for the contribution :slightly_smiling_face: drop your username if you'd like to receive the Contributor chat badge

ZemyCode commented:
@AnatoleAM thanks! If anyone reports any bugs with this feature I will look into them. I have already received the contributor badge from the "copy" button addition 😄 .

AnatoleAM commented:
Oh, right, I forgor :skull:
